internal/core/src/exec/HashTable.cpp

```diff
@@ -0,0 +1,334 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "HashTable.h"
+#include <memory>
+#include "common/SimdUtil.h"
+#include "exec/operator/OperatorUtils.h"
+
+namespace milvus {
+namespace exec {
+
+void
+populateLookupRows(const TargetBitmapView& activeRows,
+                   std::vector<vector_size_t>& lookupRows) {
+    if (activeRows.all()) {
+        std::iota(lookupRows.begin(), lookupRows.end(), 0);
+    } else {
+        auto start = -1;
+        lookupRows.clear();
+        lookupRows.reserve(activeRows.count());
+        do {
+            auto next_active = activeRows.find_next(start);
+            if (!next_active.has_value())
+                break;
+            auto next_active_row = next_active.value();
+            lookupRows.emplace_back(next_active_row);
+            start = next_active_row;
+        } while (true);
+    }
+}
+
+void
+BaseHashTable::prepareForGroupProbe(HashLookup& lookup,
+                                    const RowVectorPtr& input,
+                                    TargetBitmap& activeRows,
+                                    bool ignoreNullKeys) {
+    auto& hashers = lookup.hashers_;
+    int numKeys = hashers.size();
+    // set up column vector to each column
+    for (auto i = 0; i < numKeys; i++) {
+        auto& hasher = hashers[i];
+        auto column_idx = hasher->ChannelIndex();
+        ColumnVectorPtr column_ptr =
+            std::dynamic_pointer_cast<ColumnVector>(input->child(column_idx));
+        AssertInfo(column_ptr != nullptr,
+                   "Failed to get column vector from row vector input");
+        hashers[i]->setColumnData(column_ptr);
+        // deselect null values
+        if (ignoreNullKeys) {
+            int64_t length = column_ptr->size();
+            TargetBitmapView valid_bits_view(column_ptr->GetValidRawData(),
+                                             length);
+            activeRows &= valid_bits_view;
+        }
+    }
+    lookup.reset(activeRows.size());
+
+    const auto mode = hashMode();
+    for (auto i = 0; i < hashers.size(); i++) {
+        if (mode == BaseHashTable::HashMode::kHash) {
+            TargetBitmapView tmp_views(activeRows);
+            hashers[i]->hash(i > 0, tmp_views, lookup.hashes_);
+        } else {
+            PanicInfo(
+                milvus::OpTypeInvalid,
+                "Not support target hashMode, only support kHash for now");
+        }
+    }
+    TargetBitmapView active_views(activeRows);
+    populateLookupRows(active_views, lookup.rows_);
+}
+
+class ProbeState {
+ public:
+    enum class Operation { kProbe, kInsert, kErase };
+    static constexpr int32_t kFullMask = 0xffff;
+
+    int32_t
+    row() const {
+        return row_;
+    }
+
```

Comment:
three steps for each hash operation: 
preProbe: found target bucket
firstProbe: load first hit bit
fullProbe: iterate the matched rows in the bucket or turn to the following buckets